### ISSUE TITLE
Debug Console: wrong order of messages, when variable references are sent via OutputEvents

### ISSUE THREAD

**Issue by bxantus** [microsoft/vscode#83674]
- VSCode Version: 1.39.2, insiders: 1.40.0
- OS Version: Windows_NT x64 10.0.17134

When multiple output events are sent out from a DebugAdapter in a quick succession (for ex. processing a bunch of console.log requests), and these events contain varaiable references, the messages will mix up.
I did a research before posting this issue. It is related to issue #33822 and #79196

Steps to Reproduce:

1. When debugging a node.js process upload the following function (from debug console):
```typescript
function consoleTest() {
	console.log("first1");
	console.log("obj1", {a: 10, arr: [20,30]});
	for (var i=0; i<10; ++i) { console.log({name: "prof" + i, size: [10 + i,20], order: {i:i} });
	   console.log("done", i)
	}
	console.log("next1");
	console.log("next2");
	console.log("next3");
}
```
2. call `consoleTest()`. Because the issue is timing related, probably you have to do it multiple times, but after some tries, wou will see that objects and message strings get messed up (on my machine around "prof5")

Does this issue occur when all extensions are disabled?: Yes

I checked out the latest version of vscode's source code and tried to identify the source of the bug.
Here's an exerpt from the source code (from `debugSession.ts`):
```typescript
let outpuPromises: Promise<void>[] = [];
this.rawListeners.push(this.raw.onDidOutput(async event => {
	// ...
	// Make sure to append output in the correct order by properly waiting on preivous promises #33822
	const waitFor = outpuPromises.slice();
	// ...
	if (event.body.variablesReference) {
		const container = new ExpressionContainer(this, undefined, event.body.variablesReference, generateUuid());
		outpuPromises.push(container.getChildren().then(async children => {
			await Promise.all(waitFor);
			children.forEach(child => {
				// Since we can not display multiple trees in a row, we are displaying these variables one after the other (ignoring their names)
				(<any>child).name = null;
				this.appendToRepl(child, outputSeverity, source);
			});
		}));
	} else if (typeof event.body.output === 'string') {
		await Promise.all(waitFor);
		this.appendToRepl(event.body.output, outputSeverity, source);
	}

	await Promise.all(outpuPromises);   // <- root cause
	outpuPromises = [];                 // <- root cause
}));
```

I belive, that the lines marked with `root cause` cause the bug. 
Let me explain through an example:
- you receive an output event with a variable reference (out1) and outputPromises array is empty at this point
- a new promise which will write the contents of the varref will be appended to outputPromises: [out1]
- the function reaches the `await Promise.all(outpuPromises)` == `await Promise.all([out1])`  line, where it's execution will be suspended
- in this time you recieve an other output event with a varref (out2), the content outPromises will change to [out1, out2]
- we reach the `await Promise.all...` line, suspend execution
- meanwhile processing out1 is finished, the function processing out1 event will resume and clears the outputPromises array. From here on the ordering of messages is not guaranteed! out2 will follow out1, but when a simple string message arrives the outputPromises array is empty and it won't wait for completion of processing the previous messages

I would recommend using just one promise instead of outPromises which you can replace in each event call, something like this:
```typescript
let outputPromise = Promise.resolve();
this.rawListeners.push(this.raw.onDidOutput(async event => {
	// ...
	// Make sure to append output in the correct order by properly waiting on preivous promises #33822
	// ...
	if (event.body.variablesReference) {
		const container = new ExpressionContainer(this, undefined, event.body.variablesReference, generateUuid());
		outputPromise = outputPromise.then(() => container.getChildren()).then(async children => {
			children.forEach(child => {
				// Since we can not display multiple trees in a row, we are displaying these variables one after the other (ignoring their names)
				(<any>child).name = null;
				this.appendToRepl(child, outputSeverity, source);
			});
		});
	} else if (typeof event.body.output === 'string') {
		await outputPromise;
		this.appendToRepl(event.body.output, outputSeverity, source);
	}
}));
```

This way it is guaranteed, that output will be in the same order as events are received, as each new event processor depends on the promise value made in the step before it (if it was a varref).

An alternative solution would be to use an array as a message queue, containing the closures which append the repl lines, and a bool when they are unblocked, after getting the varref contents, the corresponding item could be unblocked, and the queue could be flushed (elements are removed from the front until they are unblocked)

I also checked the pull request proposed by dgozman: #81403. I don't think it will solve this issue (but maybe I'm wrong).
Hope this helps.

**Comment by isidorn** [comment]
@bxantus Thank you very much for this nice analysis. Sorry for the slow response, I was on vacation.
Would you be willing to create a PR that tackles this? More about how to contribute can be found here https://github.com/microsoft/vscode/wiki/How-to-Contribute

Also fyi @dgozman since he tackled a similar issue as you mentioned

**Comment by dgozman** [comment]
@bxantus There are multiple problems currently, with #81403 fixing another one (you can see the repro case in the test). The one you are analyzing here is indeed a separate issue, which I tackled [here](https://github.com/microsoft/vscode/pull/81196/files#diff-6f7452867991ca30ede51b3f40cf1aa7) in a way similar to your alternative proposed solution - unfortunately that was never merged.

**Comment by bxantus** [comment]
@isidorn I considered to create a PR, but unfortunately I didn't had the time for it (if I make a PR, i would like to try it out before, maybe write some unit tests, etc.)

Let me know, if you have the time to integrate the `patch`. If not, I will try to create a PR.

**Comment by isidorn** [comment]
@bxantus I will not integrate the patch. You can create a PR when you find the time, there is no hurry here. Thanks

**Comment by bxantus** [comment]
I checked the code recently, and saw that @isidorn solved this with fed69adea3fccae29fb04f6a0144da4acad9f8f0. Thanks!

Closing the issue.

**Comment by isidorn** [comment]
Thanks for letting us know this is good now.
